### PR TITLE
`sage.graphs`: More modularization

### PR DESCRIPTION
**Pull request by mkoeppe**
### :books: Description

Mostly using `lazy_import` and adding `# optional` annotations to doctests.

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
The linter failures are not coming from the PR.

File: src/sage/graphs/dot2tex_utils.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def quoted_latex(x):
 
     EXAMPLES::
 
-        sage: sage.graphs.dot2tex_utils.quoted_latex(matrix([[1,1],[0,1],[0,0]]))
+        sage: sage.graphs.dot2tex_utils.quoted_latex(matrix([[1,1],[0,1],[0,0]]))     # optional - sage.modules
```
Comment (dcoudert) [inline src/sage/graphs/dot2tex_utils.py]: in this file, the alignement of `# optional - sage.modules` is different than in other files

File: src/sage/graphs/generic_graph.py [hunk 2]
```diff
@@ -19490,8 +19538,8 @@ def latex_options(self):
             sage: opts = g.latex_options()
             sage: opts
             LaTeX options for Petersen graph: {}
-            sage: opts.set_option('tkz_style', 'Classic')
-            sage: opts
+            sage: opts.set_option('tkz_style', 'Classic')                                       # optional - sage.plot
```
Comment (dcoudert) [inline src/sage/graphs/generic_graph.py]: different alignment

File: src/sage/graphs/generic_graph.py [hunk 3]
```diff
@@ -19519,9 +19567,9 @@ def set_latex_options(self, **kwds):
         EXAMPLES::
 
             sage: g = graphs.PetersenGraph()
-            sage: g.set_latex_options(tkz_style='Welsh')
-            sage: opts = g.latex_options()
-            sage: opts.get_option('tkz_style')
+            sage: g.set_latex_options(tkz_style='Welsh')                                        # optional - sage.plot
```
Comment (dcoudert) [inline src/sage/graphs/generic_graph.py]: different alignment

File: src/sage/graphs/generic_graph.py [hunk 4]
```diff
@@ -23231,13 +23284,13 @@ def is_vertex_transitive(self, partition=None, verbosity=0,
             sage: G.is_vertex_transitive()
             False
             sage: P = graphs.PetersenGraph()
-            sage: P.is_vertex_transitive()
+            sage: P.is_vertex_transitive()                                     # optional - sage.groups
```
Comment (dcoudert) [inline src/sage/graphs/generic_graph.py]: different alignment

File: src/sage/graphs/generic_graph.py [hunk 5]
```diff
@@ -23231,13 +23284,13 @@ def is_vertex_transitive(self, partition=None, verbosity=0,
...
             True
             sage: D = graphs.DodecahedralGraph()
-            sage: D.is_vertex_transitive()
+            sage: D.is_vertex_transitive()                                     # optional - sage.groups
             True
-            sage: R = graphs.RandomGNP(2000, .01)
-            sage: R.is_vertex_transitive()
+            sage: R = graphs.RandomGNP(2000, .01)                              # optional - networkx
```
Comment (dcoudert) [inline src/sage/graphs/generic_graph.py]: here too

File: src/sage/graphs/graph_decompositions/vertex_separation.pyx [hunk 6]
```diff
@@ -954,8 +954,8 @@ def vertex_separation_exp(G, verbose=False):
     Graphs with non-integer vertices::
 
         sage: from sage.graphs.graph_decompositions.vertex_separation import vertex_separation_exp
-        sage: D=digraphs.DeBruijn(2,3)
-        sage: vertex_separation_exp(D)
+        sage: D = digraphs.DeBruijn(2,3)                                         # optional - sage.combinat
```
Comment (dcoudert) [inline src/sage/graphs/graph_decompositions/vertex_separation.pyx]: different alignement

File: src/sage/graphs/graph_decompositions/vertex_separation.pyx [hunk 7]
```diff
@@ -1211,12 +1211,12 @@ def width_of_path_decomposition(G, L):
     Path decomposition of a BalancedTree::
 
         sage: from sage.graphs.graph_decompositions import vertex_separation
-        sage: G = graphs.BalancedTree(3,2)
-        sage: pw, L = vertex_separation.path_decomposition(G)
-        sage: pw == vertex_separation.width_of_path_decomposition(G, L)
+        sage: G = graphs.BalancedTree(3,2)                                       # optional - networkx
```
Comment (dcoudert) [inline src/sage/graphs/graph_decompositions/vertex_separation.pyx]: different alignement

File: src/sage/graphs/graph_decompositions/vertex_separation.pyx [hunk 8]
```diff
@@ -1305,9 +1305,9 @@ def _vertex_separation_MILP_formulation(G, integrality=False, solver=None):
     EXAMPLES::
 
         sage: from sage.graphs.graph_decompositions.vertex_separation import _vertex_separation_MILP_formulation
-        sage: G = digraphs.DeBruijn(2,3)
-        sage: p, x, u, y, z = _vertex_separation_MILP_formulation(G)
-        sage: p
+        sage: G = digraphs.DeBruijn(2,3)                                         # optional - sage.combinat
```
Comment (dcoudert) [inline src/sage/graphs/graph_decompositions/vertex_separation.pyx]: different alignement

File: src/sage/graphs/graph_decompositions/vertex_separation.pyx [hunk 9]
```diff
@@ -1420,12 +1420,12 @@ def vertex_separation_MILP(G, integrality=False, solver=None, verbose=0,
     Vertex separation of a De Bruijn digraph::
 
         sage: from sage.graphs.graph_decompositions import vertex_separation
-        sage: G = digraphs.DeBruijn(2,3)
-        sage: vs, L = vertex_separation.vertex_separation_MILP(G); vs
+        sage: G = digraphs.DeBruijn(2,3)                                         # optional - sage.combinat
```
Comment (dcoudert) [inline src/sage/graphs/graph_decompositions/vertex_separation.pyx]: different alignement

**Review by dcoudert — CHANGES_REQUESTED** [review]
I noted a few places in which the alignement of the `# optional ...` is different than in the other files. Otherwise, this looks good to me. I also checked the examples you changed and this seems ok.

File: src/sage/graphs/generic_graph.py (hunk 3 above)
Comment (mkoeppe) [reply src/sage/graphs/generic_graph.py]: Thanks for spotting these! Fixed now.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**